# TWAS
a) Manhattan plot of glyco-TWAS results (unadjusted data); annotate points to indicate MR significance



## prepare data
TWAS and MR results generaated by pipelines were combined using the following commands
```
python /mnt/lustre/home/rl3328/rl3328/motor_qtl/combine_twas.py /mnt/lustre/lab/ctcn/hklein/motor_qtl_project/unadjusted_wp_2020/susie_twas/noQC_nonimputed_result/twas --type both --output ROSMAP_eQTL_wp_unadjusted
```
### Input
Input files are all on the new AWS-based HPC
* Combined TWAS file:`/mnt/lustre/lab/ctcn/hklein/motor_qtl_project/wp_2020/susie_twas/noQC_nonimputed_result/twas/ROSMAP_eQTL_wp_unadjusted.combined_twas.tsv.gz`
* Combined MR file:`/mnt/lustre/lab/ctcn/hklein/motor_qtl_project/unadjusted_wp_2020/susie_twas/noQC_nonimputed_result/twas/ROSMAP_eQTL_wp_unadjusted.combined_mr.tsv.gz`
* reference file for gene annotation: `/mnt/lustre/home/rl3328/rl3328/resource/Homo_sapiens.GRCh38.103.chr.reformatted.collapse_only.gene.region_list.2025`


### Landscape

* The TWAS results show substantial variation in the number of available genes across contexts. Monocyte, DLPFC, and AC have notably fewer genes available for analysis. In particular, the DLPFC bulk eQTL dataset includes only 58 genes, which likely explains why no genes reach significance under the most accurate predictive model for DLPFC.
* Consistently, the MR analysis also yields no significant genes for monocyte, DLPFC, or AC.

* **Given these limitations, it may be more appropriate to combine bulk and single-nucleus (sn) data for DLPFC in downstream analyses.**



#### TWAS

In [131]:
library(data.table)
library(tidyverse)
twas <- fread('/mnt/lustre/lab/ctcn/hklein/motor_qtl_project/unadjusted_wp_2020/susie_twas/noQC_nonimputed_result/twas/ROSMAP_eQTL_wp_unadjusted.combined_twas.tsv.gz')

In [132]:
head(twas)
dim(twas)

chr molecular_id    TSS       start    end       context         
1 10  ENSG00000055950 100987515 99320000 102120000 Inh_DeJager_eQTL
2 10  ENSG00000055950 100987515 99320000 102120000 Inh_DeJager_eQTL
3 10  ENSG00000055950 100987515 99320000 102120000 Inh_DeJager_eQTL
4 10  ENSG00000055950 100987515 99320000 102120000 Inh_DeJager_eQTL
5 10  ENSG00000055950 100987515 99320000 102120000 PCC_DeJager_eQTL
6 10  ENSG00000055950 100987515 99320000 102120000 PCC_DeJager_eQTL
  gwas_study         method is_imputable is_selected_method rsq_cv       
1 unadjusted_wp_2020 enet   TRUE         FALSE              -0.0002636756
2 unadjusted_wp_2020 lasso  TRUE         FALSE              -0.0008220066
3 unadjusted_wp_2020 mrash  TRUE         FALSE               0.0074837805
4 unadjusted_wp_2020 susie  TRUE          TRUE               0.0157468862
5 unadjusted_wp_2020 enet   TRUE         FALSE               0.1176354683
6 unadjusted_wp_2020 lasso  TRUE         FALSE               0.1157138583
  pval_cv      twas_z     twas_pval  type block                    
1 3.460743e-01 -0.8321645 0.40531612 eQTL chr10_100331627_104378781
2 4.181955e-01 -0.9241237 0.35542192 eQTL chr10_100331627_104378781
3 4.222085e-02 -2.4786840 0.01318681 eQTL chr10_100331627_104378781
4 5.809734e-03 -0.6523170 0.51419672 eQTL chr10_100331627_104378781
5 7.716102e-14 -0.2702251 0.78698710 eQTL chr10_100331627_104378781
6 1.253604e-13 -0.1913938 0.84821705 eQTL chr10_100331627_104378781
  method_selected_original region                    study_context 
1 FALSE                    chr10_100331627_104378781 ROSMAP_eQTL_wp
2 FALSE                    chr10_100331627_104378781 ROSMAP_eQTL_wp
3 FALSE                    chr10_100331627_104378781 ROSMAP_eQTL_wp
4  TRUE                    chr10_100331627_104378781 ROSMAP_eQTL_wp
5 FALSE                    chr10_100331627_104378781 ROSMAP_eQTL_wp
6 FALSE                    chr10_100331627_104378781 ROSMAP_eQTL_wp
  source_file                                                    
1 ROSMAP_eQTL_wp_unadjusted.chr10_100331627_104378781.twas.tsv.gz
2 ROSMAP_eQTL_wp_unadjusted.chr10_100331627_104378781.twas.tsv.gz
3 ROSMAP_eQTL_wp_unadjusted.chr10_100331627_104378781.twas.tsv.gz
4 ROSMAP_eQTL_wp_unadjusted.chr10_100331627_104378781.twas.tsv.gz
5 ROSMAP_eQTL_wp_unadjusted.chr10_100331627_104378781.twas.tsv.gz
6 ROSMAP_eQTL_wp_unadjusted.chr10_100331627_104378781.twas.tsv.gz

[1] 168334     20

In [ ]:
## TWAS analysis were performed across 10 contexts and 13872 genes
twas %>% pull(context) |> unique()#%>%length

twas %>% pull(molecular_id)%>%unique()%>%length



[1] "Inh_DeJager_eQTL"     "PCC_DeJager_eQTL"     "Mic_DeJager_eQTL"    
 [4] "Exc_DeJager_eQTL"     "Oli_DeJager_eQTL"     "Ast_DeJager_eQTL"    
 [7] "OPC_DeJager_eQTL"     "AC_DeJager_eQTL"      "DLPFC_DeJager_eQTL"  
[10] "monocyte_ROSMAP_eQTL"

[1] 13872

In [175]:
13872-60-3031-3

[1] 10778

In [ ]:
## 13872 total genes, 10778 are from DLPFC
twas |> group_by(context) |> summarize(gene = n_distinct(molecular_id))

context              gene 
1  AC_DeJager_eQTL         60
2  Ast_DeJager_eQTL      5854
3  DLPFC_DeJager_eQTL      57
4  Exc_DeJager_eQTL      8187
5  Inh_DeJager_eQTL      6377
6  Mic_DeJager_eQTL      2729
7  OPC_DeJager_eQTL      3031
8  Oli_DeJager_eQTL      5392
9  PCC_DeJager_eQTL     10376
10 monocyte_ROSMAP_eQTL     3

In [135]:

twas_res_sig <- twas %>% filter(is_selected_method=="TRUE")%>%filter(twas_pval<=0.05)#%>%pull(context)%>%unique()
twas_res_sig|> pull(molecular_id)%>%unique()%>%length


[1] 4325

In [136]:
twas_DLPFC = twas |> filter(context == 'DLPFC_DeJager_eQTL')
dim(twas_DLPFC)

[1] 232  20

In [ ]:
## Best model-0 gene significant after bonferroni correction
twas |> filter(is_selected_method, twas_pval < unadj_cutoff, str_detect(context, "DLPFC"))
## All models-1 gene(EPHA3) significant after bonferroni correction
twas |> filter(twas_pval < unadj_cutoff, str_detect(context, "DLPFC"))


chr molecular_id TSS start end context gwas_study method is_imputable
     is_selected_method rsq_cv pval_cv twas_z twas_pval type block
     method_selected_original region study_context source_file

chr molecular_id    TSS      start    end      context           
1 3   ENSG00000044524 89107621 86600000 93720000 DLPFC_DeJager_eQTL
  gwas_study         method is_imputable is_selected_method rsq_cv    
1 unadjusted_wp_2020 mrash  TRUE         FALSE              0.04285215
  pval_cv      twas_z    twas_pval    type block                 
1 2.935038e-09 -5.764569 8.186698e-09 eQTL chr3_88209300_94537986
  method_selected_original region                 study_context 
1 FALSE                    chr3_88209300_94537986 ROSMAP_eQTL_wp
  source_file                                                 
1 ROSMAP_eQTL_wp_unadjusted.chr3_88209300_94537986.twas.tsv.gz

#### MR

In [49]:
mr <- fread('/mnt/lustre/lab/ctcn/hklein/motor_qtl_project/unadjusted_wp_2020/susie_twas/noQC_nonimputed_result/twas/ROSMAP_eQTL_wp_unadjusted.combined_mr.tsv.gz')

In [50]:
mr_sig_unadj <- mr %>% .[
  meta_pval < 0.05 / .N &          # Bonferroni correction
  cpip >= 0.7 &                    # strong causal evidence
  num_CS >= 1 &                    # as quested by Hans and Gao on Oct 24, 2025 to get a loose version
  Q_pval > 0.01 &                  # no heterogeneity
  I2 < 0.4                         # low heterogeneity
]

In [138]:
mr_sig_unadj |> filter(gene_name == 'ENSG00000044524')

Q Q_pval I2 context cpip gene_name gwas_study meta_eff meta_pval num_CS
     num_IV se_meta_eff region study_context source_file

In [139]:
head(mr_sig_unadj)
dim(mr_sig_unadj)

Q     Q_pval I2 context          cpip  gene_name       gwas_study        
1 1.823 0.402  0  Exc_DeJager_eQTL 0.983 ENSG00000138111 unadjusted_wp_2020
2 0.000 1.000  0  Inh_DeJager_eQTL 0.956 ENSG00000138111 unadjusted_wp_2020
3 0.000 1.000  0  Ast_DeJager_eQTL 1.012 ENSG00000171206 unadjusted_wp_2020
4 0.001 0.979  0  Exc_DeJager_eQTL 1.026 ENSG00000171206 unadjusted_wp_2020
5 0.000 1.000  0  Exc_DeJager_eQTL 0.960 ENSG00000078403 unadjusted_wp_2020
6 0.000 1.000  0  PCC_DeJager_eQTL 0.995 ENSG00000148572 unadjusted_wp_2020
  meta_eff meta_pval num_CS num_IV se_meta_eff region                   
1  0.000   0         3       21    0           chr10_100331627_104378781
2  0.000   0         1       16    0           chr10_100331627_104378781
3  0.001   0         1       59    0           chr10_100331627_104378781
4  0.001   0         2      144    0           chr10_100331627_104378781
5 -0.001   0         1        8    0           chr10_20625737_23053090  
6 -0.002   0         1       92    0           chr10_62446953_64035328  
  study_context 
1 ROSMAP_eQTL_wp
2 ROSMAP_eQTL_wp
3 ROSMAP_eQTL_wp
4 ROSMAP_eQTL_wp
5 ROSMAP_eQTL_wp
6 ROSMAP_eQTL_wp
  source_file                                                         
1 ROSMAP_eQTL_wp_unadjusted.chr10_100331627_104378781.mr_result.tsv.gz
2 ROSMAP_eQTL_wp_unadjusted.chr10_100331627_104378781.mr_result.tsv.gz
3 ROSMAP_eQTL_wp_unadjusted.chr10_100331627_104378781.mr_result.tsv.gz
4 ROSMAP_eQTL_wp_unadjusted.chr10_100331627_104378781.mr_result.tsv.gz
5 ROSMAP_eQTL_wp_unadjusted.chr10_20625737_23053090.mr_result.tsv.gz  
6 ROSMAP_eQTL_wp_unadjusted.chr10_62446953_64035328.mr_result.tsv.gz

[1] 206  15

In [140]:
mr_sig_unadj |> group_by(context) |> summarize(gene = n_distinct(gene_name))

context          gene
1 Ast_DeJager_eQTL 21  
2 Exc_DeJager_eQTL 41  
3 Inh_DeJager_eQTL 29  
4 Mic_DeJager_eQTL  6  
5 OPC_DeJager_eQTL  7  
6 Oli_DeJager_eQTL 27  
7 PCC_DeJager_eQTL 75

### summarize significant TWAS & MR genes

* Significant TWAS associations were defined using two criteria: (i) the most accurate prediction model yielded a TWAS p-value below the Bonferroni-corrected threshold, or (ii) at least half of the prediction models produced p-values below this threshold.
* Among the 13,872 total genes, 10,778 are from DLPFC. Overall, 126 genes show significant TWAS associations, of which 80 are significant in DLPFC tissues

* Mendelian randomization (MR) associations were considered significant if they met all of the following criteria: a meta-analysis p-value passing Bonferroni correction (meta-analysis p < 0.05 divided by the number of tested genes), strong causal evidence as indicated by a causal posterior inclusion probability (CPIP) ≥ 0.7, at least one credible set (num_CS ≥ 1), and no evidence of substantial heterogeneity (Cochran’s Q test p > 0.01 and I² < 0.4).
* 206 genes show significant MR associations, of which 131 are significant in DLPFC tissues.

In [141]:
n_imputable_genes <- twas %>%
  filter(is_imputable) %>%                  # keep only imputable rows
  distinct(molecular_id, context, gwas_study) %>%         # count unique gene-context pairs
  nrow()
n_imputable_genes

[1] 42066

In [151]:
# Compute cutoff using Bonferroni-style correction
unadj_cutoff <- 0.05 / n_imputable_genes
unadj_cutoff

[1] 1.188608e-06

In [171]:
twas_unadj_sig <- twas %>%
  mutate(sig = twas_pval < unadj_cutoff) %>%
  group_by(molecular_id, context, gwas_study) %>%
  summarise(
    n_methods = n(),
    n_sig = sum(sig, na.rm = T),
    best_sig = any(sig & is_selected_method, na.rm = T),
    .groups = "drop"
  ) %>%
  mutate(pass = (n_sig / n_methods >= 0.5) | best_sig) %>%
  mutate(pass_strict = (n_sig / n_methods >= 0.5)) %>%
  filter(pass) %>% left_join(twas) %>% arrange(twas_pval) %>% distinct(molecular_id, .keep_all = T)
twas_unadj_sig %>% dim

Joining with `by = join_by(molecular_id, context, gwas_study)`


[1] 126  25

In [178]:
twas_unadj_sig |> filter(!context %in% c("monocyte_ROSMAP_eQTL", "PCC_DeJager_eQTL", "AC_DeJager_eQTL")) |> nrow()

[1] 80

In [154]:
gene_ref <- fread('/mnt/lustre/home/rl3328/rl3328/resource/Homo_sapiens.GRCh38.103.chr.reformatted.collapse_only.gene.region_list.2025')

In [180]:
# twas_unadj_sig |> left_join(gene_ref, by = c("molecular_id" = "gene_id")) |> select(molecular_id, gene_name, everything())

## Figure data

In [181]:
df <- twas %>%
  mutate(sig = twas_pval < unadj_cutoff) %>%
  group_by(molecular_id, context, gwas_study) %>%
  summarise(
    n_methods = n(),
    n_sig = sum(sig, na.rm = TRUE),
    best_pval = min(twas_pval, na.rm = TRUE),
    best_method_pval = min(twas_pval[is_selected_method], na.rm = TRUE),
    best_method_selected = any(sig & is_selected_method, na.rm = TRUE),
    .groups = "drop"
  ) %>%
  mutate(
    pass = (n_sig / n_methods >= 0.5) | best_method_selected, # pass defined by two threshold, best or half methods satisfied
    final_pval = ifelse(n_sig / n_methods >= 0.5, best_pval, best_method_pval),
    # fallback: if no selected method existed (Inf), use best_pval
    final_pval = ifelse(is.infinite(final_pval), best_pval, final_pval)
  ) %>%
  left_join(twas, by = c("molecular_id", "context", "gwas_study")) %>%
  filter(twas_pval == final_pval) %>%
  distinct(molecular_id, .keep_all = TRUE) %>%
  mutate(
    logp = -log10(twas_pval) * sign(twas_z),
    pos = (start + end) / 2
  ) %>%
  merge(gene_ref, by.x = 'molecular_id', by.y = 'gene_id')

# Order chromosomes
df$chr <- factor(df$chr, levels = as.character(1:22))

Warning message:
“There were 211 warnings in `summarise()`.
The first warning was:
ℹ In argument: `best_method_pval = min(twas_pval[is_selected_method], na.rm =
  TRUE)`.
ℹ In group 202: `molecular_id = "ENSG00000005189"`, `context =
  "Exc_DeJager_eQTL"`, `gwas_study = "unadjusted_wp_2020"`.
Caused by warning in `min()`:
! no non-missing arguments to min; returning Inf
ℹ Run `dplyr::last_dplyr_warnings()` to see the 210 remaining warnings.”


In [182]:
dim(df)

[1] 13872    33

In [183]:
head(df)

molecular_id    context          gwas_study         n_methods n_sig
1 ENSG00000000419 Exc_DeJager_eQTL unadjusted_wp_2020 4         0    
2 ENSG00000000457 Ast_DeJager_eQTL unadjusted_wp_2020 4         0    
3 ENSG00000000460 Ast_DeJager_eQTL unadjusted_wp_2020 4         0    
4 ENSG00000000971 PCC_DeJager_eQTL unadjusted_wp_2020 4         0    
5 ENSG00000001036 Ast_DeJager_eQTL unadjusted_wp_2020 4         0    
6 ENSG00000001084 Ast_DeJager_eQTL unadjusted_wp_2020 4         0    
  best_pval   best_method_pval best_method_selected pass  final_pval  ⋯
1 0.004887676 0.634741658      FALSE                FALSE 0.634741658 ⋯
2 0.314317590 0.666951750      FALSE                FALSE 0.666951750 ⋯
3 0.550799973 0.681781965      FALSE                FALSE 0.681781965 ⋯
4 0.006727627 0.006727627      FALSE                FALSE 0.006727627 ⋯
5 0.125012981 0.125012981      FALSE                FALSE 0.125012981 ⋯
6 0.166978326 0.851335412      FALSE                FALSE 0.851335412 ⋯
  method_selected_original region                   study_context 
1 TRUE                     chr20_50319421_53436693  ROSMAP_eQTL_wp
2 TRUE                     chr1_168438717_170228106 ROSMAP_eQTL_wp
3 TRUE                     chr1_168438717_170228106 ROSMAP_eQTL_wp
4 TRUE                     chr1_195599253_199271134 ROSMAP_eQTL_wp
5 TRUE                     chr6_143399926_145632539 ROSMAP_eQTL_wp
6 TRUE                     chr6_52730905_54027603   ROSMAP_eQTL_wp
  source_file                                                    logp      
1 ROSMAP_eQTL_wp_unadjusted.chr20_50319421_53436693.twas.tsv.gz  -0.1974030
2 ROSMAP_eQTL_wp_unadjusted.chr1_168438717_170228106.twas.tsv.gz  0.1759056
3 ROSMAP_eQTL_wp_unadjusted.chr1_168438717_170228106.twas.tsv.gz -0.1663545
4 ROSMAP_eQTL_wp_unadjusted.chr1_195599253_199271134.twas.tsv.gz  2.1721381
5 ROSMAP_eQTL_wp_unadjusted.chr6_143399926_145632539.twas.tsv.gz -0.9030449
6 ROSMAP_eQTL_wp_unadjusted.chr6_52730905_54027603.twas.tsv.gz   -0.0698993
  pos       #chr  TSS.y     TES       gene_name
1  51747434 chr20  50958555  50934867 DPM1     
2 169540000 chr1  169894267 169849631 SCYL3    
3 169540000 chr1  169662007 169854080 C1orf112 
4 195093752 chr1  196652043 196747504 CFH      
5 143115860 chr6  143511720 143494812 FUCA2    
6  53557156 chr6   53616970  53497341 GCLC

In [184]:
chr_info <- df %>%
  group_by(chr) %>%
  summarise(chr_len = max(pos)) %>%
  mutate(chr_start = lag(cumsum(chr_len), default = 0))

df <- df %>%
  left_join(chr_info, by = "chr") %>%
  mutate(cum_pos = pos + chr_start)
  
axis_df <- chr_info %>%
  mutate(center = chr_start + chr_len/2)


In [185]:
df <- mr_sig_unadj %>% rename('molecular_id' = 'gene_name') %>% merge(df, . , all.x = T)

In [186]:
df %>% filter(!is.na(Q_pval)) %>% select(molecular_id, context, gene_name)

molecular_id context gene_name

In [187]:

# Add a column for MR significance (example: FDR < 0.05)
df <- df %>%
  mutate(category = case_when(
    !is.na(Q_pval) & pass ~ "TWAS & MR",        # Both significant
    is.na(Q_pval) & pass ~ "TWAS only",         # Only TWAS significant
    !pass & is.na(Q_pval) ~ "Not Significant",  # Neither significant
    TRUE ~ "Others"                              # Catch-all for remaining cases
  ))

In [188]:
# dir.create('data')
saveRDS(list(df = df, unadj_cutoff = unadj_cutoff, axis_df= axis_df), 'data/TWAS_manhattan_plot_data.rds')